### ISSUE TITLE
Diffing public API against a published crate with `[package] name` != `[lib] name` miscalculates the rustdoc JSON path and fails

### ISSUE THREAD

**Issue by V0ldek** [cargo-public-api/cargo-public-api#561]
This is similar to an earlier issue I reported (and that was fixed): #476 

This time I don't have an easy repro, because it seems to require the crate to be actually published on crates.io. You could nonetheless repro it with the project I'm working on: https://github.com/V0ldek/rsonpath

## Bug

Diffing against a previous version of the crate doesn't work if used to check a library on a package with a different name.

The root cause appears to be similar to #476, only that this time the rustdoc file gets cached into `$HOME/.cache/cargo-public-api/build-root-for-published-crates`.

## Repro

1. Clone `rsonpath` and go to the library crate:

```console
> git clone https://github.com/V0ldek/rsonpath.git && cd ./rsonpath/crates/rsonpath-lib
```

**Note**: if my hypothesis is correct this should work with any package that produces a library with a different name and was previously published to crates.io.

2. Try to diff against a published version, for example 0.8.6:

```console
> cargo public-api diff 0.8.6
   Compiling proc-macro2 v1.0.79
   Compiling unicode-ident v1.0.12
   Compiling libc v0.2.153
   Compiling getrandom v0.1.16
    Checking cfg-if v1.0.0
   Compiling syn v1.0.109
    Checking ppv-lite86 v0.2.17
   Compiling thiserror v1.0.58
    Checking minimal-lexical v0.2.1
    Checking memchr v2.7.2
    Checking unicode-width v0.1.11
    Checking log v0.4.21
    Checking smallvec v1.13.2
    Checking static_assertions v1.1.0
    Checking nom v7.1.3
   Compiling quote v1.0.35
   Compiling syn v2.0.55
    Checking memmap2 v0.9.4
    Checking rand_core v0.5.1
    Checking rand_chacha v0.2.2
    Checking rand v0.7.3
   Compiling thiserror-impl v1.0.58
    Checking rsonpath-syntax v0.2.0
   Compiling contracts v0.4.0
    Checking vector-map v1.0.1
 Documenting rsonpath-lib v0.8.6
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 2.22s
Error: Failed to parse rustdoc JSON at "/home/mat/.cache/cargo-public-api/build-root-for-published-crates/rsonpath-lib/-/0.8.6/target/doc/rsonpath_lib.json".

This version of `cargo public-api` requires at least:

    nightly-2023-08-25

Ensure your nightly toolchain is up to date with:

    rustup install nightly --profile minimal

If that does not help, it might be `cargo public-api` that is out of date. Try
to install the latest version with

    cargo install cargo-public-api --locked

If the issue remains, please report at

    https://github.com/Enselic/cargo-public-api/issues

Caused by:
    No such file or directory (os error 2)
```

**Note**: after the last step, the `.cache/cargo-public-api/build-root-for-published-crates/rsonpath-lib/-/0.8.6/target/doc` directory does contain a file named `rsonpath.json`, because the `rsonpath-lib` package produces a library called `rsonpath`.

## Expected behaviour

cargo-public-api should recognise the library target name and look for the correct rustdoc file in the cache.

## Additional details

Rust version: stable, 1.76.0
Rust nightly version: rustc 1.79.0-nightly (c9f8f3438 2024-03-27)
cargo-public-api version: 0.34.0

**Comment by V0ldek** [comment]
Probably of note: this is a relatively minor issue, because it only involves diffing against crates.io. A workaround is to simply use the diff against commits locally, which is easy if you tag your releases consistently.

In this case it's simply:
```console
cargo public-api diff v0.8.6..HEAD
```

**Comment by Enselic** [comment]
Thanks for reporting. I was able to reproduce using your provided steps.

The problem seems to be that we use a fake Cargo.toml to trick cargo into downloading dependencies (including dependencies of dependencies) for us. And then we rely on that rustdoc by default builds rustdoc JSON also for dependencies. But we don't actually parse the Cargo.toml of dependencies.

For example, with the reproducer, cargo-public-api uses this manifest to figure out the JSON path: `/home/martin/.cache/cargo-public-api/build-root-for-published-crates/rsonpath-lib/-/0.8.6/Cargo.toml`

This manifest looks like this:

```toml
[package]
name = "crate-downloader"
version = "0.1.0"
edition = "2021"

[lib]
path = "lib.rs"

[dependencies.rsonpath-lib]
version = "=0.8.6"
```

and is constructed here:

https://github.com/Enselic/cargo-public-api/blob/0b77df9bfd3e93c6d4fbba275d38ba05645f36e6/cargo-public-api/src/published_crate.rs#L106-L113

And from this info it concludes that it should find the rustdoc JSON at `/home/martin/.cache/cargo-public-api/build-root-for-published-crates/rsonpath-lib/-/0.8.6/target/doc/rsonpath_lib.json` since the `[lib]` section does not contain a `name` field.

One way that we could solve this could be to actually open `~/.cargo/registry/cache/index.crates.io-6f17d22bba15001f/rsonpath-lib-0.8.6.crate` and parse its Cargo.toml so we can figure out the right library name and thus rustdoc JSON path.

Not sure if I can bring myself to do this, but I would be happy to mentor anyone who wants to try :) Might be a bit tricky to find the path to the dependency on all platforms in all environments... But it's probably quite easy to make it work in most cases, and we can always try to fallback on the current mechanism if we don't find the downloaded .crate.

**Comment by Enselic** [comment]
I have published our test-crate https://crates.io/crates/other-lib-name so this is now very easy to reproduce. To build and reproduce with the code in the git repo, you simply run this command:

```console
$ cargo run -- --manifest-path test-apis/other-lib-name/Cargo.toml diff 0.1.0
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/cargo-public-api --manifest-path test-apis/other-lib-name/Cargo.toml diff 0.1.0`
 Documenting other-lib-name v0.1.0
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.22s
Error: Failed to parse rustdoc JSON at "/home/martin/.cache/cargo-public-api/build-root-for-published-crates/other-lib-name/-/0.1.0/target/doc/other_lib_name.json".

This version of `cargo public-api` requires at least:

    nightly-2023-08-25

Ensure your nightly toolchain is up to date with:

    rustup install nightly --profile minimal

If that does not help, it might be `cargo public-api` that is out of date. Try
to install the latest version with

    cargo install cargo-public-api --locked

If the issue remains, please report at

    https://github.com/Enselic/cargo-public-api/issues

Caused by:
    No such file or directory (os error 2)
```

**Edit:** I think I am close to having a fix, I just need some more time...

**Comment by Enselic** [comment]
This wasn't so tricky to fix after all. This should be fixed by https://github.com/Enselic/cargo-public-api/pull/562. At least it fixes the regression test I added.
